Source/Core/DolphinQt/Achievements/AchievementProgressWidget.cpp

```diff
@@ -49,6 +51,41 @@ AchievementProgressWidget::AchievementProgressWidget(QWidget* parent) : QWidget(
 QGroupBox*
 AchievementProgressWidget::CreateAchievementBox(const rc_api_achievement_definition_t* achievement)
 {
+  QLabel* a_badge = new QLabel();
+  const auto unlock_status = AchievementManager::GetInstance()->GetUnlockStatus(achievement->id);
+  bool is_unlocked =
```

Comment:
I wonder if this would be a bit more readable if the boolean was broken down.  Also, is hardcode mode expected to always return an `UnlockType::HARDCORE`?  Feels odd to have both checks.

Reply:
So there's hardcore mode, and "softcore" mode (which is just ... not hardcore mode). When an achievement is unlocked in softcore mode, it's still locked in hardcore mode, but if it's unlocked in hardcore mode it's unlocked in both. So what this check is doing is, if the achievement is (remotely) unlocked in hardcore it's always unlocked, or if it's unlocked in softcore AND the manager is currently in softcore, it's marked as unlocked. If the achievement is unlocked in softcore but not hardcore mode, and the manager is currently in hardcore mode, it hasn't been unlocked yet.

Unsure quite how to break it down to be more readable, do you have a recommendation?

Reply:
The only way I could see this being a bit better is to switch the logic a bit.  Right now you check all the states to determine if it's unlocked to determine the badge type.  Then you do it _again_ effectively to determine the color.  You could just do it once?

Example:

```
AchievementManager::BadgeStatus* badge = unlock_status.locked_badge;
std::string_view color = AchievementManager::GRAY;
if (unlock_status.remote_unlock_status == AchievementManager::UnlockStatus::UnlockType::HARDCOR)
{
  color = AchievementManager::GOLD;
  badge = &unlock_status.unlocked_badge;
}
else if (unlock_status.remote_unlock_status == AchievementManager::UnlockStatus::UnlockType::SOFTCORE)
{
  color = AchievementManager::BLUE;
  badge = &unlock_status.unlocked_badge;
}
else if (...)
{
}

// Now set the pixel data, set the badge visible, and set the style sheet
```

Reply:
Ahhh, I love changes like this, incredibly easy to code and test, but an absolute _pain in the ass_ to rebase onto the existing commits in a way that makes organizational sense ^^;